File: Source/LefticeRuntime/Private/Core/ConsoleVariable.cpp
// Copyright (c) NextTurn.
// See LICENSE.txt in the project root for more information.

#include "LefticeRuntime/Private/Exports.hpp"

#include "LefticeRuntime/Private/MarshaledTypes.hpp"
#include "LefticeRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"
#include "Runtime/Core/Public/HAL/ConsoleManager.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME ConsoleVariable

TYPE_EXPORT_START

int32_t EXPORT_CALL_CONV GetInt32(const IConsoleVariable& Variable)
{
	return Variable.GetInt();
}

float EXPORT_CALL_CONV GetSingle(const IConsoleVariable& Variable)
{
	return Variable.GetFloat();
}

void EXPORT_CALL_CONV GetString(const IConsoleVariable& Variable, FMarshaledString& Result)
{
	Result = Variable.GetString();
}

void EXPORT_CALL_CONV SetInt32(IConsoleVariable& Variable, int32_t Value)
{
	Variable.Set(Value);
}

void EXPORT_CALL_CONV SetSingle(IConsoleVariable& Variable, float Value)
{
	Variable.Set(Value);
}

void EXPORT_CALL_CONV SetString(IConsoleVariable& Variable, const char16_t* Value)
{
	Variable.Set(CHAR16_TO_TCHAR(Value));
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(GetInt32),
	EXPORT_METHOD(GetSingle),
	EXPORT_METHOD(GetString),
	EXPORT_METHOD(SetInt32),
	EXPORT_METHOD(SetSingle),
	EXPORT_METHOD(SetString),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

File: Managed/Leftice.Runtime/Core/ConsoleVariable.cs
// Copyright (c) NextTurn.
// See LICENSE.txt in the project root for more information.

using System;
using Leftice.Processors;

namespace Unreal
{
    public sealed class ConsoleVariable : ConsoleObject
    {
        internal ConsoleVariable(IntPtr pointer) : base(pointer) { }

        public int ValueAsInt32
        {
            get => NativeMethods.GetInt32(this.Pointer);
            set => NativeMethods.SetInt32(this.Pointer, value);
        }

        public float ValueAsSingle
        {
            get => NativeMethods.GetSingle(this.Pointer);
            set => NativeMethods.SetSingle(this.Pointer, value);
        }

        public unsafe string ValueAsString
        {
            get
            {
                NativeMethods.GetString(this.Pointer, out ScriptArray nativeResult);
                return Leftice.StringMarshaler.ToManagedFinally(&nativeResult);
            }

            set => NativeMethods.SetString(this.Pointer, value);
        }

        private static class NativeMethods
        {
            [Calli]
            public static extern int GetInt32(IntPtr variable);

            [Calli]
            public static extern float GetSingle(IntPtr variable);

            [Calli]
            public static extern void GetString(IntPtr variable, out ScriptArray nativeResult);

            [Calli]
            public static extern void SetInt32(IntPtr variable, int value);

            [Calli]
            public static extern void SetSingle(IntPtr variable, float value);

            [Calli]
            public static extern void SetString(IntPtr variable, string value);
        }
    }
}
